File: python/src/rivu_server_sdk/state_ui.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, NonNegativeInt, PositiveInt, model_validator


class UiMountV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    messageId: str = Field(min_length=1)
    slot: str = Field(min_length=1)
    order: int


class UiComponentErrorV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    code: str = Field(min_length=1)
    message: str = Field(min_length=1)
    details: dict[str, Any] | None = None


class UiComponentV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    type: str = Field(min_length=1)
    schemaVersion: PositiveInt
    props: dict[str, Any]
    state: dict[str, Any] | None = None
    revision: NonNegativeInt
    mounts: list[UiMountV1]
    status: Literal["building", "ready", "error"] | None = None
    error: UiComponentErrorV1 | None = None

    @model_validator(mode="after")
    def _validate_lifecycle(self) -> "UiComponentV1":
        status = self.status or "ready"
        if status == "error":
            if self.error is None:
                raise ValueError("component.error must be set when status=error")
            return self
        if self.error is not None:
            raise ValueError("component.error must be omitted unless status=error")
        return self


class UiStateV1(BaseModel):
    model_config = ConfigDict(extra="allow")

    v: Literal[1]
    components: dict[str, UiComponentV1]

    @model_validator(mode="after")
    def _validate_component_ids(self) -> "UiStateV1":
        for component_id in self.components.keys():
            if not component_id.strip():
                raise ValueError("componentId must be non-empty")
        return self

File: packages/rivu-ui-spec/src/state-ui.ts
import { z } from 'zod';

const nonEmptyString = z.string().min(1);
const nonNegativeInt = z.number().int().min(0);
const positiveInt = z.number().int().min(1);

export const uiMountV1Schema = z
  .object({
    messageId: nonEmptyString,
    slot: nonEmptyString,
    order: z.number().int(),
  })
  .strict();

export type UiMountV1 = z.output<typeof uiMountV1Schema>;

export const uiComponentErrorV1Schema = z
  .object({
    code: nonEmptyString,
    message: nonEmptyString,
    details: z.record(z.string(), z.unknown()).optional(),
  })
  .passthrough();

export type UiComponentErrorV1 = z.output<typeof uiComponentErrorV1Schema>;

export const uiComponentV1Schema = z
  .object({
    type: nonEmptyString,
    schemaVersion: positiveInt,
    props: z.record(z.string(), z.unknown()),
    state: z.record(z.string(), z.unknown()).optional(),
    revision: nonNegativeInt,
    mounts: z.array(uiMountV1Schema),
    status: z.enum(['building', 'ready', 'error']).optional(),
    error: uiComponentErrorV1Schema.optional(),
  })
  .passthrough()
  .superRefine((component, ctx) => {
    const status = component.status ?? 'ready';
    if (status === 'error') {
      if (!component.error) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'component.error must be set when status=error',
        });
      }
      return;
    }
    if (component.error) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'component.error must be omitted unless status=error',
      });
    }
  });

export type UiComponentV1 = z.output<typeof uiComponentV1Schema>;

export const uiStateV1Schema = z
  .object({
    v: z.literal(1),
    components: z
      .record(z.string(), uiComponentV1Schema)
      .superRefine((components, ctx) => {
        for (const id of Object.keys(components)) {
          if (!id.trim()) {
            ctx.addIssue({
              code: z.ZodIssueCode.custom,
              message: 'componentId must be non-empty',
            });
            return;
          }
        }
      }),
  })
  .passthrough();

export type UiStateV1 = z.output<typeof uiStateV1Schema>;
